juju_verify/verifiers/neutron_gateway.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2021 Canonical Limited.
+#
+# This file is part of juju-verify.
+#
+# juju-verify is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# juju-verify is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program. If not, see https://www.gnu.org/licenses/.
+"""neutron-gateway verification."""
+import json
+from typing import List
+
+from juju.unit import Unit
+from packaging.version import Version
+
+from juju_verify.verifiers.base import BaseVerifier, Result, Severity
+from juju_verify.verifiers.result import aggregate_results
+from juju_verify.utils.action import data_from_action
+from juju_verify.utils.unit import parse_charm_name, run_action_on_unit
+
+
+class NeutronGateway(BaseVerifier):
+    """Implementation of verification checks for the neutron-gateway charm."""
+
+    NAME = 'neutron-gateway'
+    action_name_result_map = {"get-status-routers": "router-list",
+                              "get-status-dhcp": "dhcp-networks",
+                              "get-status-lb": "load-balancers"}
+    action_name_failure_string_map = {"get-status-routers": ("The following routers are "
+                                                             "non-redundant: {}"),
+                                      "get-status-dhcp": ("The following DHCP networks "
+                                                          "are non-redundant: {}"),
+                                      "get-status-lb": ("The following LBaasV2 LBs were "
+                                                        "found: {}. LBaasV2 does not "
+                                                        "offer HA.")}
+
+    @classmethod
+    def get_unit_resource_list(cls, unit: Unit,
+                               get_resource_action_name: str) -> List[dict]:
+        """Given a get resource action, return the relevant resources on the unit."""
+        get_resource_action = run_action_on_unit(unit, get_resource_action_name)
+        action_name_res = cls.action_name_result_map[get_resource_action_name]
+        resource_list_json = data_from_action(get_resource_action, action_name_res)
+        resource_list = json.loads(resource_list_json)
+        return resource_list
+
+    def get_all_ngw_units(self) -> List[Unit]:
+        """Get all neutron-gateway units, including those not being shutdown."""
+        return [unit for unit in self.model.units.values()
+                if parse_charm_name(unit.data.get('charm-url', '')) == self.NAME]
```

Comment:
It seems to me this is a generic helper that could be used in many verifiers. I think the Ceph verifiers also use a similar approach.

Reply:
@mkalcok Can you report a bug for this and then resolve this conversation.

Reply:
[New bug opened](https://bugs.launchpad.net/juju-verify/+bug/1937262)